clients/privacy-center/packages/fides-consent/src/lib/consent-banner.ts

```diff
@@ -0,0 +1,553 @@
+    try {
+      // eslint-disable-next-line no-new
+      new URL(options.privacyCenterUrl);
+    } catch (e) {
+      debugLog("Invalid banner options: geolocationApiUrl is an invalid URL!", options);
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Navigates to the Fides Privacy Center to manage consent settings
+ */
+const navigateToPrivacyCenter = (): void => {
+  const options: ConsentBannerOptions = getBannerOptions();
+  debugLog("Navigate to Privacy Center URL:", options.privacyCenterUrl);
+  if (options.privacyCenterUrl) {
+    window.location.assign(options.privacyCenterUrl);
+  }
+}
+
+/**
+ * Fetch the user's geolocation from an external API
+ */
+const getLocation = async (): Promise<UserGeolocation> => {
+  debugLog("Running getLocation...");
+  const options = getBannerOptions();
+  if (!options.geolocationApiUrl) {
+    debugLog("Missing geolocationApiUrl, cannot get location!");
+    return {};
+  }
+
+  debugLog(`Calling geolocation API: GET ${options.geolocationApiUrl}...`);
+  const fetchOptions: RequestInit = {
+    mode: "cors"
+  };
+  const response = await fetch(options.geolocationApiUrl,fetchOptions);
+
+  if (!response.ok) {
+    debugLog("Error getting location from geolocation API, returning {}. Response:", response);
+    return {};
+  }
+
+  try {
+    const body = await response.json();
+    debugLog("Got location response from geolocation API, returning:", body);
+    return body;
+  } catch (e) {
+    debugLog("Error parsing response body from geolocation API, returning {}. Response:", response);
+    return {};
+  }
+}
+
+/**
+ * Determine whether or not the banner should be enabled for the given location
+ */
+const isBannerEnabledForLocation = (location?: UserGeolocation): boolean => {
+  const options = getBannerOptions();
+  if (location === undefined || !location) {
+    debugLog("Location unknown, assume banner must be shown. isBannerEnabledForLocation = true");
+    return true;
+  }
+
+  // Get the user's country
+  if (location.country === undefined || !location.country) {
+    debugLog("Country unknown, assume banner must be shown. isBannerEnabledForLocation = true");
+    return true;
+  }
+
+  if (options.isEnabledCountries) {
+    if (options.isEnabledCountries.includes(location.country)) {
+      debugLog(`Country ${location.country} included in isEnabledCountries, banner must be shown. isBannerEnabledForLocation = true`);
+      return true;
+    }
+    debugLog(`Country ${location.country} not included in isEnabledCountries, banner must be hidden. isBannerEnabledForLocation = false`);
+    return false;
+  }
+
+  debugLog("No location-specific rules matched, assume banner must be shown. isBannerEnabledForLocation = true");
+  return true;
+}
+
+/**
+ * Builds a button DOM element with the given id, class name, and text label
+ */
+const buildButton = (id: string, className: string, label?: string, onClick?: (event: MouseEvent) => void): HTMLButtonElement => {
```

Comment:
The next couple functions (`buildButton`, `buildBanner`, `buildStyles`) are lots of imperative DOM APIs and probably should be refactored into a separate file like `consent-banner-dom.ts`